src/web-components.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/**
+ * @type {boolean|undefined}
+ * @visibleForTesting
+ */
+let shadowDomSupported;
+
+/**
+ * @type {Function|undefined}
+ * @visibleForTesting
+ */
+let shadowDomMethod;
+
+let win = window;
```

Comment:
Unfortunately we can't have window-dependent side effects in our source code.  Further `getShadowDomMethod().call(...)` might have issues if resolved via `win.Element` because we also allow most of APIs to be called from friendly iframes.

What I'm thinking instead, maybe we could try to have a string var: `shadowDomSupport`. It could have values `none`, `v0` and `v1`. This way we can set this value from tests and try different code paths.

In that case, `isShadowDomSupported() {return shadowDomSupport != 'none'}`. And we could switch `getShadowDomMethod()` to something like this:

```
createShadowRoot(host) {
  switch(getShadowDomSupport()) {
     case 'v0':
        return host.createShadowRoot();
     case 'v1':
        return host.attachShadow({...});
  }
  assert(...);
}
```